File: backend/src/main/java/org/akashihi/mdg/service/AccountService.java
package org.akashihi.mdg.service;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.akashihi.mdg.api.v1.RestException;
import org.akashihi.mdg.dao.AccountRepository;
import org.akashihi.mdg.dao.CategoryRepository;
import org.akashihi.mdg.dao.CurrencyRepository;
import org.akashihi.mdg.dao.OperationRepository;
import org.akashihi.mdg.entity.Account;
import org.akashihi.mdg.entity.AccountType;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

import javax.transaction.Transactional;
import java.math.BigDecimal;
import java.util.Collection;
import java.util.Map;
import java.util.Optional;

import static org.akashihi.mdg.dao.AccountSpecification.filteredAccount;

@Service
@RequiredArgsConstructor
@Slf4j
public class AccountService {
    private final AccountRepository accountRepository;
    private final BudgetService budgetService;
    private final CategoryRepository categoryRepository;
    private final CurrencyRepository currencyRepository;
    private final TransactionService transactionService;
    private final RateService rateService;
    private final OperationRepository operationRepository;

    protected Account applyBalance(Account a) {
        var balance = accountRepository.getBalance(a.getId()).orElse(BigDecimal.ZERO);
        a.setBalance(balance);
        a.setPrimaryBalance(rateService.toCurrentDefaultCurrency(a.getCurrency(), a.getBalance()));
        return a;
    }

    @Transactional
    public Account create(Account account) {
        if (!account.getAccountType().equals(AccountType.ASSET)) {
            if (account.getOperational() != null && account.getOperational()) {
                throw new RestException("ACCOUNT_NONASSET_INVALIDFLAG", 412, "/accounts");
            }
            if (account.getFavorite() != null && account.getFavorite()) {
                throw new RestException("ACCOUNT_NONASSET_INVALIDFLAG", 412, "/accounts");
            }
        }
        if (account.getAccountType().equals(AccountType.ASSET) && account.getCategoryId() == null) { //Default category for asset accounts
                var defaultCategory = categoryRepository.findByNameAndAccountType("Current", AccountType.ASSET).orElseThrow(() ->new RestException("CATEGORY_NOT_FOUND", 404, "/accounts"));
                account.setCategory(defaultCategory);
                account.setCategoryId(defaultCategory.getId());
        }
        var currency = currencyRepository.findById(account.getCurrencyId()).orElseThrow(() ->new RestException("CURRENCY_NOT_FOUND", 404, "/accounts"));
        account.setCurrency(currency);
        if (account.getCategoryId() != null) {
            var category = categoryRepository.findById(account.getCategoryId()).orElseThrow(() ->new RestException("CATEGORY_NOT_FOUND", 404, "/accounts"));
            if (!category.getAccountType().equals(account.getAccountType())) {
                throw new RestException("CATEGORY_INVALID_TYPE", 412, "/accounts");
            }
            account.setCategory(category);
        }
        account.setHidden(false);
        accountRepository.save(account);
        log.info("Created account {}", account);
        return applyBalance(account);
    }

    @Transactional
    public Collection<Account> list(Optional<Map<String, String>> query) {
        var sort = Sort.by("accountType").ascending().and(Sort.by("name").ascending());
        if (query.isEmpty()) {
            return accountRepository.findAll(sort).stream().map(this::applyBalance).toList();
        }
        return accountRepository.findAll(filteredAccount(query.get()), sort).stream().map(this::applyBalance).toList();
    }

    @Transactional
    public Collection<Account> listByType(AccountType type) {
        return accountRepository.findAllByAccountType(type).stream().map(this::applyBalance).toList();
    }

    @Transactional
    public Optional<Account> get(Long id) {
        return accountRepository.findById(id).map(this::applyBalance);
    }

    @Transactional
    public Optional<Account> update(Long id, Account newAccount) {
        var accountValue = accountRepository.findById(id);
        if (accountValue.isEmpty()) {
            return accountValue;
        }

        var account = accountValue.get();
        if (newAccount.getHidden() != null) {
            account.setHidden(newAccount.getHidden());
        }
        if (newAccount.getName() != null) {
            account.setName(newAccount.getName());
        }
        if (newAccount.getCategoryId() == null) {
            account.setCategory(null);
        } else {
            Long currentCategoryId = null;
            if (account.getCategory() != null) {
                currentCategoryId = account.getCategory().getId();
            }
            if (!newAccount.getCategoryId().equals(currentCategoryId)) {
                var newCategory = categoryRepository.findById(newAccount.getCategoryId()).orElseThrow(() ->new RestException("CATEGORY_NOT_FOUND", 404, "/accounts/%d".formatted(id)));
                if (!newCategory.getAccountType().equals(account.getAccountType())) {
                    throw new RestException("CATEGORY_INVALID_TYPE", 412, "/accounts/%d".formatted(id));
                }
                account.setCategory(newCategory);
            }
        }

        if (account.getAccountType() == AccountType.ASSET) {
            account.setFavorite(newAccount.getFavorite());
            account.setOperational(newAccount.getOperational());
            if (!account.getCurrency().getId().equals(newAccount.getCurrencyId())) {
                throw new RestException("ACCOUNT_CURRENCY_ASSET", 422, "/accounts/%d".formatted(id));
            }
        } else {
            if (!account.getCurrency().getId().equals(newAccount.getCurrencyId())) {
                var currencyValue = currencyRepository.findById(newAccount.getCurrencyId());
                currencyValue.ifPresent(currency -> transactionService.updateTransactionsCurrencyForAccount(account, currency));
                currencyValue.ifPresent(currency -> budgetService.updateCurrencyForAccount(account, currency));
                currencyValue.ifPresent(account::setCurrency);
            }
            if (newAccount.getFavorite() != null && newAccount.getFavorite() || newAccount.getOperational() != null && newAccount.getOperational()) {
                throw new RestException("ACCOUNT_NONASSET_INVALIDFLAG", 412, "/accounts/%d".formatted(id));
            }
        }
        accountRepository.save(account);
        return Optional.of(applyBalance(account));
    }

    @Transactional
    public void delete(Long id) {
        var account = accountRepository.findById(id).orElseThrow(() -> new RestException("ACCOUNT_NOT_FOUND", 404, "/accounts/%d".formatted(id)));
        if (!isDeletable(account.getId())) {
            throw new RestException("ACCOUNT_IN_USE", 409, "/accounts/%d".formatted(id));
        }
        accountRepository.delete(account);
    }

    @Transactional public Boolean isDeletable(Long id) {
        var account = accountRepository.findById(id).orElseThrow(() -> new RestException("ACCOUNT_NOT_FOUND", 404, "/accounts/%d/status".formatted(id)));
        return !operationRepository.doOperationsExistForAccount(account.getId()).orElse(false);
    }
}

File: backend/src/main/kotlin/org/akashihi/mdg/service/AccountService.kt
package org.akashihi.mdg.service

import org.akashihi.mdg.api.v1.MdgException
import org.akashihi.mdg.dao.AccountRepository
import org.akashihi.mdg.dao.AccountSpecification
import org.akashihi.mdg.dao.CategoryRepository
import org.akashihi.mdg.dao.CurrencyRepository
import org.akashihi.mdg.dao.OperationRepository
import org.akashihi.mdg.entity.Account
import org.akashihi.mdg.entity.AccountType
import org.slf4j.LoggerFactory
import org.springframework.data.domain.Sort
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*
import javax.persistence.EntityManager
import javax.persistence.PersistenceContext

@Service
open class AccountService(private val accountRepository: AccountRepository, private val budgetService: BudgetService, private val categoryRepository: CategoryRepository, private val currencyRepository: CurrencyRepository, private val transactionService: TransactionService, private val operationRepository: OperationRepository,  @PersistenceContext private val em: EntityManager) {
    @Transactional
    open fun create(account: Account): Account {
        if (account.accountType != AccountType.ASSET) {
            if (account.operational == true) {
                throw MdgException("ACCOUNT_NONASSET_INVALIDFLAG")
            }
            if (account.favorite == true) {
                throw MdgException("ACCOUNT_NONASSET_INVALIDFLAG")
            }
        }
        if (account.accountType == AccountType.ASSET && account.categoryId == null) { // Default category for asset accounts
            val defaultCategory = categoryRepository.findByNameAndAccountType("Current", AccountType.ASSET) ?: throw MdgException("CATEGORY_NOT_FOUND")
            account.category = defaultCategory
            account.categoryId = defaultCategory.id
        }
        val currency = currencyRepository.findByIdOrNull(account.currencyId) ?: throw MdgException("CURRENCY_NOT_FOUND")
        account.currency = currency
        account.categoryId?.also {
            val category = categoryRepository.findByIdOrNull(account.categoryId) ?: throw MdgException("CATEGORY_NOT_FOUND")
            if (category.accountType != account.accountType) {
                throw MdgException("CATEGORY_INVALID_TYPE")
            }
            account.category = category
        }
        account.hidden = false
        accountRepository.save(account)
        log.info("Created account {}", account)
        return account
    }

    @Transactional(readOnly = true)
    open fun list(query: Map<String, String>): Collection<Account> {
        val sort = Sort.by("accountType").ascending().and(Sort.by("name").ascending())
        return if (query.isEmpty()) {
            accountRepository.findAll(sort)
        } else {
            accountRepository.findAll(AccountSpecification.filteredAccount(query), sort)
        }
    }

    @Transactional(readOnly = true)
    open fun listByType(type: AccountType): Collection<Account> = accountRepository.findAllByAccountType(type)

    @Transactional(readOnly = true)
    open fun listPopularByType(type: AccountType): Collection<Account> = accountRepository.getPopularAccountsForType(type.toDbValue()).mapNotNull { this.get(it) }

    @Transactional(readOnly = true)
    open operator fun get(id: Long): Account? = accountRepository.findByIdOrNull(id)

    @Transactional
    open fun update(id: Long, newAccount: Account): Account? {
        val account = accountRepository.findByIdOrNull(id) ?: return null
        if (newAccount.hidden != null) {
            account.hidden = newAccount.hidden
        }
        if (newAccount.name != null) {
            account.name = newAccount.name
        }
        if (newAccount.categoryId == null) {
            account.category = null
        } else {
            val currentCategoryId = account.category?.id
            if (newAccount.categoryId != currentCategoryId) {
                val newCategory = categoryRepository.findByIdOrNull(newAccount.categoryId) ?: throw MdgException("CATEGORY_NOT_FOUND")
                if (newCategory.accountType != account.accountType) {
                    throw MdgException("CATEGORY_INVALID_TYPE")
                }
                account.category = newCategory
            }
        }
        if (account.accountType === AccountType.ASSET) {
            account.favorite = newAccount.favorite
            account.operational = newAccount.operational
            if (account.currency!!.id != newAccount.currencyId) {
                throw MdgException("ACCOUNT_CURRENCY_ASSET")
            }
        } else {
            if (account.currency!!.id != newAccount.currencyId) {
                val currencyValue = currencyRepository.findByIdOrNull(newAccount.currencyId)
                currencyValue?.also { transactionService.updateTransactionsCurrencyForAccount(account, it) }
                currencyValue?.also { budgetService.updateCurrencyForAccount(account, it) }
                currencyValue?.also { account.currency = it }
            }
            if (newAccount.favorite == true || newAccount.operational == true) {
                throw MdgException("ACCOUNT_NONASSET_INVALIDFLAG")
            }
        }
        accountRepository.saveAndFlush(account)
        em.refresh(account)
        return account
    }

    @Transactional
    open fun delete(id: Long) {
        val account = accountRepository.findByIdOrNull(id) ?: throw MdgException("ACCOUNT_NOT_FOUND")
        if (!isDeletable(account.id)) {
            throw MdgException("ACCOUNT_IN_USE")
        }
        accountRepository.delete(account)
    }

    @Transactional(readOnly = true)
    open fun isDeletable(id: Long?): Boolean {
        val account = accountRepository.findByIdOrNull(id) ?: throw MdgException("ACCOUNT_NOT_FOUND")
        return !(operationRepository.doOperationsExistForAccount(account.id!!) ?: false)
    }

    companion object {
        @Suppress("JAVA_CLASS_ON_COMPANION")
        @JvmStatic
        private val log = LoggerFactory.getLogger(javaClass.enclosingClass)
    }
}
